### DISCUSSION TITLE
The python wrappers now reflect the Delphi inheritance hierarchy

### DISCUSSION THREAD

**Discussion by pyscripter** [pyscripter/python4delphi#399] [Announcements]
With the fix to #382, tp_base is correctly set for wrappers descending from TPyDelphiObject and registered through the standard RegisterDelphiWrapper mechanism.  This means that the isinstance and issubclass builtins will work correctly on wrapped Delphi objects and classes.

It also means that we should not call inherited in RegisterMethods and RegisterGetSets, since python will find the inherited methods and getsets anyway.  Calling inherited does not do any harm (it overrides the methods in inherited classes), but it is inefficient speed and memory-wise.

Wrapped classes which do not provide a RegisterMethods and RegisterGetSets override, they should provide an empty one.  Otherwise the inherited ones would be created on both the parent and the child class leading to unnecessary duplication.

https://github.com/pyscripter/python4delphi/commit/45ca26c8ddea451af6279c4ff65c0e53fe251e22 removed the calls to inherited from existing  implementations of RegisterMethods and RegisterGetSets.  It also changed the implementation of the `__dir__`  wrapper. However, now,  empty RegisterMethods and RegisterGetSets, should be added to the wrappers that do not have these methods.

@lmbelo, @jimmckeeth

**Comment by lmbelo** [comment]
We have this fixed and many other enhancements in the latest "docs" feature branch.

Have a look: https://github.com/Embarcadero/python4delphi/tree/docs

1) All methods and properties provided by Delphi are now built-in;
2) It follows the Delphi inheritance until Python object;
3) Python is now able to refer to class attributes;
4) The dir method now lists all Delphi attributes available in a class (not instances), categorized by class parents.
5) Callables are not created on-the-fly anymore;
7) Doc strings are being read from XML files;
8) IDE's can use type attributes as stubs for code completion, as they are built-in now;
8) and more...

@jimmckeeth @pyscripter

**Reply by pyscripter** [reply]
Sounds great!  I will have a close look.

But we should coordinate, so what I do here does not conflict with what you do at the Embarcadero fork.

**Reply by lmbelo** [reply]
Agree!

**Reply by pyscripter** [reply]
> 3\. Python is now able to refer to class attributes;

Class methods static or otherwise have been supported for a long time.  See for instance https://github.com/pyscripter/python4delphi/blob/master/Tutorials/Webinar%20I/WrapDelphiDemo/MainForm.pas#L49

What is not supported is class fields and properties, but AFAIK (see https://en.delphipraxis.net/topic/8087-access-class-fields-and-properties-with-rtti/) this is not currently possible.   

@lmbelo Could you please clarify what is the new functionality in that respect and how it was implemented.

**Reply by lmbelo** [reply]
When referred from a given instance of a type, not the type itself, right?

**Reply by pyscripter** [reply]
Yes.  I see.  Now you can call class methods from the type?

**Reply by lmbelo** [reply]
Yes, exactly. Class and static methods from a type.

**Reply by pyscripter** [reply]
Good!
